manage_sections/templates/manage_sections/create_section_form.html

```diff
@@ -435,41 +435,46 @@ <h4 class="modal-title" id="confirmDelSectionLabel">Confirm Delete</h4>
         }
     }
 
-    $("#sectionsMenu").on('click', '.remove_data', function(e) {
-        var url = this.href;
-        var userLine = $(this).closest("li");
-        var enrollmentCount = parseInt(userLine.find(".sectionCount").text());
-
-        addEnrollmentCountText(enrollmentCount)
-        $('#confirmDelSection').modal({ keyboard:false }) 
-        .one('click', '#btnConfirmDelSection', function() {
-            $.ajax({
-                url: url,
-                type : "POST",
-                success : function(json) {
-                    $("#message").hide();
-                    userLine.fadeOut('slow', function(){
-                        userLine.remove();
-                    });
-                    show_if_none();
-                },
-                error : function(xhr) {
-                    $("#message").addClass('alert alert-danger').show();
-                    if (xhr.status ==422){
-                         $("#message").text(xhr.responseText);
-                    }
-                    else{
-                        $("#message").text("Error: Section has not been deleted ");
-                    }
+    function onClickRemoveSection(e) {
+        e.preventDefault(); // do not allow browser to load the href URL
+        const currentSectionRow = $(this).closest('li');
+        $('#btnConfirmDelSection').data({url: this.href, sectionId: currentSectionRow.data('sectionId')});
+        const enrollmentCount = parseInt(currentSectionRow.find(".sectionCount").text());
+        addEnrollmentCountText(enrollmentCount);
+        $('#confirmDelSection').modal({ keyboard:false }).show();
+    }
+
+    function onClickConfirmRemoveSection(e) {
+        const removeUrl = $(this).data('url');
+        const sectionId = $(this).data('sectionId');
+        const sectionRow = $('[data-section-id="' + sectionId + '"]');
+        $.ajax({
+            url: removeUrl,
+            type : "POST",
+            success : function(json) {
+                $("#message").hide();
+                sectionRow.fadeOut('slow', function() {
+                    sectionRow.remove();
+                });
+                show_if_none();
+            },
+            error : function(xhr) {
+                $("#message").addClass('alert alert-danger').show();
+                if (xhr.status == 422) {
+                        $("#message").text(xhr.responseText);
+                }
+                else {
+                    $("#message").text("Error: Section has not been deleted ");
```

Comment:
Ah, so was the issue (at least in part) solved by removing the nested callbacks (lines 438-463), putting the logic into separate functions, and calling them sequentially?

Reply:
@PatrickMTonne -- kinda! The problem was that the nesting was causing more and more `click` events to be attached to the action of activating the `#btnConfirmDelSection` button. Every time that someone [clicked on the trash icon](https://github.com/Harvard-University-iCommons/canvas_manage_course/blob/v2.2.4/manage_sections/templates/manage_sections/create_section_form.html#L428) we were [adding a click listener to the confirm button](https://github.com/Harvard-University-iCommons/canvas_manage_course/blob/v2.2.4/manage_sections/templates/manage_sections/create_section_form.html#L433), but never removing it later, so events just stacked up and got fired off upon the first actual "confirm".

In other locations of our code, those events were being removed upon the modal disappearing; see for instance [this chunk of code](https://github.com/Harvard-University-iCommons/canvas_manage_course/blob/10bbfcfa3ef2bf82377f96e6032eab2999b085ec/manage_sections/static/manage_sections/js/section_details.js#L179-L181) to remove users from the sections. This section deletion code though was [missing a similar `off()` call](https://github.com/Harvard-University-iCommons/canvas_manage_course/blob/10bbfcfa3ef2bf82377f96e6032eab2999b085ec/manage_sections/static/manage_sections/js/section_details.js#L179-L181).

I wondered if maybe it was more straightforward, however, to not continuously add and remove event handlers. Instead, as Chris suggested in the ticket, events trigger an update to the modal so it can track state. In a modern framework like React state would be tracked in JS (in closures or state management), but I still think using the [HTML data attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) is a pretty standard and comprehensible pattern.